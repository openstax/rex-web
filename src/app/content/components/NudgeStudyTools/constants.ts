export const arrowDesktopHeight = 11.5;
export const arrowDesktopWidth = 15;
export const arrowMobileHeight = 6.2;
export const arrowMobileWidth = 5.2;
export const arrowRightMargin = 1;
export const arrowTopMargin = 1.6;

export const closeButtonSize = 4;
export const closeButtonDistanceFromContent = 1.6;
export const closeButtonMobileMargin = 3;

export const contentMarginTop = 1.6;
export const contentMarginLeft = 1;
export const contentWidth = 69;

export const mobileSpotlightPadding = .6;

export const cookieNudge = {
  // How many times nuge was shown
  counter: 'nudge_study_guides_counter',
  // When nudge was shown
  date: 'nudge_study_guides_date',
  // How many pages user has opened (opening the same page twice counts it as 2)
  pageCounter: 'nudge_study_guides_page_counter',
};

// Show nudge only for users which opened at least 2 pages
export const nudgeStudyToolsMinPageLimit = 2;

export const daysUntilCookieExpires = 365 * 20;

// Show nudge every 4 weeks: 4 * 7 days * 24 hours * 60 minutes * 60 seconds * 1000 ms
export const timeIntervalBetweenShowingNudgeInMs = 4 * 7 * 24 * 60 * 60 * 1000;
// Show nudge max 3 times
export const nudgeStudyToolsShowLimit = 3;
// Id of the element which should be highlighted
export const nudgeStudyToolsTargetId = 'nudge-study-tools';
export const mobileNudgeStudyToolsTargetId = 'mobile-menu-button';
