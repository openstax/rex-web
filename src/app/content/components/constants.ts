import theme from '../../theme';
import { BookWithOSWebData } from '../types';

export const bookBannerDesktopBigHeight = 13;
export const bookBannerDesktopMiniHeight = 7;
export const bookBannerMobileBigHeight = 10.4;
export const bookBannerMobileMiniHeight = 6;

export const scrollOffset = 3;

export const sidebarDesktopWidth = 37.5;
export const sidebarMobileWidth = 30.4;
export const sidebarTransitionTime = 300;

export const searchResultsBarDesktopWidth = 37.5;
export const searchResultsBarMobileWidth = 33.8;

export const toolbarIconColor = theme.color.primary.gray;

export const mobileSearchContainerMargin = 1;
export const toolbarWidth = 8;
export const toolbarSearchInputHeight = 3.2;
export const toolbarSearchInputMobileHeight = 3.2;
export const toolbarSearchInputMobileWidth = 18.5;
export const toolbarToggleSearchBarTextMobileWidth = 11.2;
export const toolbarButtonMargin = 2;
export const toolbarHrHeight = 0.1;
export const sidebarDesktopWithToolbarWidth = searchResultsBarDesktopWidth + toolbarWidth;
export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
  + (mobileSearchContainerMargin * 2)
  + toolbarHrHeight;

export const topbarDesktopHeight = 5;
export const topbarMobileHeight = 4;
export const toolbarMobileExpandedHeight = topbarMobileHeight
  + toolbarMobileSearchWrapperHeight;

export const verticalNavbarMaxWidth = 8;

export const searchSidebarTopOffset = bookBannerMobileMiniHeight
  + toolbarMobileExpandedHeight;

export const contentTextWidth = 82.5;

export const mainContentBackground = '#fff';

export const maxContentGutter = 6;
export const contentWrapperMaxWidth = contentTextWidth + sidebarDesktopWidth + verticalNavbarMaxWidth;

export const defaultTheme = 'blue' as BookWithOSWebData['theme'];
