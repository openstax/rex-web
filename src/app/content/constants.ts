import { HighlightColorEnum } from '@openstax/highlighter/dist/api';

export const studyGuidesFeatureFlag = 'studyGuidesEnabled';
export const practiceQuestionsFeatureFlag = 'practiceQuestionsEnabled';

export const maxHighlightsApiPageSize = 200;
export const maxResourcesPerFetch = 10;

export const summaryPageSize = 20;
export const loadMoreDistanceFromBottom = 50;

export const modalQueryParameterName = 'modal';
export const colorFilterQueryParameterName = 'colors';
export const locationIdsFilterQueryParameterName = 'locationIds';

export const highlightStyles: Array<{label: HighlightColorEnum, passive: string, focused: string}> = [
  {label: HighlightColorEnum.Yellow, passive: '#ffff8a', focused: '#fed200'},
  {label: HighlightColorEnum.Green, passive: '#def99f', focused: '#92d101'},
  {label: HighlightColorEnum.Blue, passive: '#c8f5ff', focused: '#00c3ed'},
  {label: HighlightColorEnum.Purple, passive: '#cbcfff', focused: '#545ec8'},
  {label: HighlightColorEnum.Pink, passive: '#ffc5e1', focused: '#de017e'},
];
