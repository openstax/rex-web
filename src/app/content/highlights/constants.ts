import { KeyCombinationOptions } from '../../reactUtils';

export const modalUrlName = 'MH';

export const cardWidth = 20;
export const cardPadding = 0.8;
export const cardContentMargin = 3;
export const cardFocusedContentMargin = 1;
export const cardMinWindowMargin = 2;
export const cardMarginBottom = 2;

export const highlightBlockPadding = 1;
export const highlightIndicatorSizeForBlock = 1.2;
export const highlightIndicatorSize = 0.9;

export const highlightKeyCombination: KeyCombinationOptions = {
  key: 'h',
};
