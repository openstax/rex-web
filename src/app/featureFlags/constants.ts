// tslint:disable:object-literal-sort-keys

import { FeatureFlagVariantValue } from './types';

export const experiments: Record<string, FeatureFlagVariantValue[]> = {
  searchButton: [
    'originalButton',
    'bannerColorButton',
    'grayButton',
  ],
  kineticBanner: [
    false,
    0,
    1,
    2,
  ],
};

export const experimentIds: { [key: string]: keyof typeof experiments } = {
  'OCCkMMCZSwW87szzpniCow': 'searchButton', // dev
  'zHtr33wcS1mVgFn7b2nU3A': 'searchButton', // prod

  'xneKbLnfQ8SbU-BvFNBM7Q': 'kineticBanner', // dev
  'L3WXXDcvT_OhGcLxyDcYOQ': 'kineticBanner', // prod
};
