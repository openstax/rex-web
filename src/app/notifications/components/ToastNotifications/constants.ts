export const clearErrorAfter = 5000;
export const shouldAutoDismissAfter = 3000;
export const fadeOutDuration = 1000;
export const timeUntilFadeIn = 5;

export const toastMessageKeys = {
  archive: {
    failure: {
      load: 'i18n:notification:toast:archive:load-failure',
    },
  },
  higlights: {
    failure: {
      create: 'i18n:notification:toast:highlights:create-failure',
      delete: 'i18n:notification:toast:highlights:delete-failure',
      load: 'i18n:notification:toast:highlights:load-failure',
      popUp: {
        load: 'i18n:notification:toast:highlights:popup:load-failure',
        print: 'i18n:notification:toast:highlights:popup:print-failure',
      },
      search: 'i18n:notification:toast:search:highlight-not-found',
      update: {
        annotation: 'i18n:notification:toast:highlights:update-failure:annotation',
        color: 'i18n:notification:toast:highlights:update-failure:color',
      },
    },
  },
  search: {
    failure: {
      nodeNotFound: 'i18n:notification:toast:search:highlight-not-found',
    },
  },
  studyGuides: {
    failure: {
      load: 'i18n:notification:toast:study-guides:load-failure',
      popUp: {
        load: 'i18n:notification:toast:study-guides:popup:load-failure',
        print: 'i18n:notification:toast:study-guides:popup:print-failure',
      },
    },
  },
};
