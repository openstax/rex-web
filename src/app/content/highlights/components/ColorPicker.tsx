import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components/macro';
import { match, not } from '../../../fpUtils';
import { highlightStyles } from '../../constants';
import { cardPadding } from '../constants';
import ColorIndicator from './ColorIndicator';

interface SingleSelectProps {
  color?: HighlightColorEnum;
  onRemove?: () => void;
  onChange: (color: HighlightColorEnum) => void;
  multiple: false | undefined;
}

interface MultipleSelectProps {
  selected: HighlightColorEnum[];
  onChange: (selected: string[]) => void;
  multiple: true;
}

type Props = (SingleSelectProps | MultipleSelectProps) & {
  className?: string;
  size?: 'small';
};

interface ColorButtonProps {
  name: string;
  checked: boolean;
  size: Props['size'];
  style: typeof highlightStyles[number];
  onClick: () => void;
  className?: string;
}

// tslint:disable-next-line:variable-name
const ColorButton = styled(({className, size, style, ...props}: ColorButtonProps) => {
  const color = useIntl().formatMessage({id: `i18n:highlighting:colors:${style.label}`});

  return <ColorIndicator
    style={style}
    size={size}
    title={color}
    aria-label={useIntl().formatMessage({id: 'i18n:highlighting:change-color'}, {color})}
    checked={props.checked}
    component={<label />}
    className={className}
  >
    <input type='checkbox' {...props} />
  </ColorIndicator>;
})`
  cursor: pointer;
  margin: 0 ${cardPadding}rem ${cardPadding}rem 0;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
  }
`;

// tslint:disable-next-line:variable-name
const ColorPicker = ({className, ...props}: Props) => {

  // its important that this be click focusable, because when clicking
  // labels in chrome the focus is first moved to a background element
  // before the input, and that might cause weird behavior in parent
  // elements.
  return <div className={className} tabIndex={-1}>
    {highlightStyles.map((style) => <ColorButton key={style.label}
      name={style.label}
      checked={props.multiple ? props.selected.includes(style.label) : props.color === style.label}
      style={style}
      size={props.size}
      onChange={() => props.multiple
        ? props.selected.includes(style.label)
          ? props.onChange(props.selected.filter(not(match(style.label))))
          : props.onChange([...props.selected, style.label])
        : props.color === style.label
          ? props.onRemove ? props.onRemove() : null
          : props.onChange(style.label)}
    />)}
  </div>;
};

export default styled(ColorPicker)`
  outline: none;
  display: flex;
  flex-direction: row;
  overflow: visible;
`;
