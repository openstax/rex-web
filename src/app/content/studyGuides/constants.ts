import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import { highlightStyles as highlightsColorStyles } from '../constants';

export const modalUrlName = 'SG';

export const highlightStyles: Array<{label: HighlightColorEnum, passive: string, focused: string}> =
  highlightsColorStyles.filter((style) => style.label !== HighlightColorEnum.Pink);

export const colorfilterLabels = new Set(highlightStyles.map(({label}) => label));
