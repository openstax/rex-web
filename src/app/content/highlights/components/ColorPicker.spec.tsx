import React from 'react';
import renderer from 'react-test-renderer';
import TestContainer from '../../../../test/TestContainer';
import { highlightStyles } from '../../constants';
import ColorPicker from './ColorPicker';

describe('ColorPicker', () => {
  it('matches snapshot no selection', () => {
    const component = renderer.create(<TestContainer>
      <ColorPicker onChange={jest.fn()} onRemove={jest.fn()} />
    </TestContainer>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('checks selection', () => {
    const component = renderer.create(<TestContainer>
      <ColorPicker color={highlightStyles[0].label} onChange={jest.fn()} onRemove={jest.fn()} />
    </TestContainer>);

    const [first, ...rest] = component.root.findAllByType('input');
    expect(first.props.checked).toEqual(true);
    rest.forEach((input) => expect(input.props.checked).toEqual(false));
  });

  it('checks selection (multiple)', () => {
    const component = renderer.create(<TestContainer>
      <ColorPicker multiple selected={[highlightStyles[0].label]} onChange={jest.fn()} onRemove={jest.fn()} />
    </TestContainer>);

    const [first, ...rest] = component.root.findAllByType('input');
    expect(first.props.checked).toEqual(true);
    rest.forEach((input) => expect(input.props.checked).toEqual(false));
  });

  it('calls update when changing selection', () => {
    const onChange = jest.fn();
    const onRemove = jest.fn();

    const component = renderer.create(<TestContainer>
      <ColorPicker color={highlightStyles[0].label} onChange={onChange} onRemove={onRemove} />
    </TestContainer>);

    const [, second] = component.root.findAllByType('input');

    second.props.onChange();

    expect(onRemove).not.toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith(highlightStyles[1].label);
  });

  it('calls update when changing selection (multiple)', () => {
    const onChange = jest.fn();
    const onRemove = jest.fn();

    const component = renderer.create(<TestContainer>
      <ColorPicker multiple selected={[highlightStyles[0].label]} onChange={onChange} onRemove={onRemove} />
    </TestContainer>);

    const [, second] = component.root.findAllByType('input');

    second.props.onChange();

    expect(onRemove).not.toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith([highlightStyles[0].label, highlightStyles[1].label]);
  });

  it('calls remove when changing selection', () => {
    const onChange = jest.fn();
    const onRemove = jest.fn();

    const component = renderer.create(<TestContainer>
      <ColorPicker color={highlightStyles[0].label} onChange={onChange} onRemove={onRemove} />
    </TestContainer>);

    const [first] = component.root.findAllByType('input');

    first.props.onChange();

    expect(onRemove).toHaveBeenCalled();
    expect(onChange).not.toHaveBeenCalled();
  });

  it('calls remove when changing selection (multiple)', () => {
    const onChange = jest.fn();
    const onRemove = jest.fn();

    const component = renderer.create(<TestContainer>
      <ColorPicker multiple selected={[highlightStyles[0].label]} onChange={onChange} onRemove={onRemove} />
    </TestContainer>);

    const [first] = component.root.findAllByType('input');

    first.props.onChange();

    expect(onRemove).not.toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith([]);
  });

  it('do not call remove if it was not passed', () => {
    const onChange = jest.fn();

    const component = renderer.create(<TestContainer>
      <ColorPicker color={highlightStyles[0].label} onChange={onChange} />
    </TestContainer>);

    const [first] = component.root.findAllByType('input');

    first.props.onChange();

    expect(onChange).not.toHaveBeenCalled();
  });
});
