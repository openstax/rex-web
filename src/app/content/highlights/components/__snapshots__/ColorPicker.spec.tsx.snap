// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker matches snapshot no selection 1`] = `
<div
  className="ColorPicker-an3xh1-1 eIpXfa"
  tabIndex={-1}
>
  <label
    aria-label="Apply yellow highlight"
    checked={false}
    className="ColorPicker__ColorButton-an3xh1-0 dcHOwJ ColorIndicator-sc-1w5zl7g-2 HsETE"
    component={<label />}
    style={
      Object {
        "focused": "#fed200",
        "label": "yellow",
        "passive": "#ffff8a",
      }
    }
    title="yellow"
  >
    <svg
      aria-hidden="true"
      className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
    <input
      checked={false}
      name="yellow"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="ColorIndicator__FocusedStyle-sc-1w5zl7g-1 gcsHaO"
      style={
        Object {
          "focused": "#fed200",
          "label": "yellow",
          "passive": "#ffff8a",
        }
      }
    />
  </label>
  <label
    aria-label="Apply green highlight"
    checked={false}
    className="ColorPicker__ColorButton-an3xh1-0 dcHOwJ ColorIndicator-sc-1w5zl7g-2 iXfiep"
    component={<label />}
    style={
      Object {
        "focused": "#92d101",
        "label": "green",
        "passive": "#def99f",
      }
    }
    title="green"
  >
    <svg
      aria-hidden="true"
      className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
    <input
      checked={false}
      name="green"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="ColorIndicator__FocusedStyle-sc-1w5zl7g-1 dmUcJc"
      style={
        Object {
          "focused": "#92d101",
          "label": "green",
          "passive": "#def99f",
        }
      }
    />
  </label>
  <label
    aria-label="Apply blue highlight"
    checked={false}
    className="ColorPicker__ColorButton-an3xh1-0 dcHOwJ ColorIndicator-sc-1w5zl7g-2 jvnIhj"
    component={<label />}
    style={
      Object {
        "focused": "#00c3ed",
        "label": "blue",
        "passive": "#c8f5ff",
      }
    }
    title="blue"
  >
    <svg
      aria-hidden="true"
      className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
    <input
      checked={false}
      name="blue"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="ColorIndicator__FocusedStyle-sc-1w5zl7g-1 hKvHax"
      style={
        Object {
          "focused": "#00c3ed",
          "label": "blue",
          "passive": "#c8f5ff",
        }
      }
    />
  </label>
  <label
    aria-label="Apply purple highlight"
    checked={false}
    className="ColorPicker__ColorButton-an3xh1-0 dcHOwJ ColorIndicator-sc-1w5zl7g-2 fEyUOh"
    component={<label />}
    style={
      Object {
        "focused": "#545ec8",
        "label": "purple",
        "passive": "#cbcfff",
      }
    }
    title="purple"
  >
    <svg
      aria-hidden="true"
      className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
    <input
      checked={false}
      name="purple"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="ColorIndicator__FocusedStyle-sc-1w5zl7g-1 pdWKj"
      style={
        Object {
          "focused": "#545ec8",
          "label": "purple",
          "passive": "#cbcfff",
        }
      }
    />
  </label>
  <label
    aria-label="Apply pink highlight"
    checked={false}
    className="ColorPicker__ColorButton-an3xh1-0 dcHOwJ ColorIndicator-sc-1w5zl7g-2 jhGpQY"
    component={<label />}
    style={
      Object {
        "focused": "#de017e",
        "label": "pink",
        "passive": "#ffc5e1",
      }
    }
    title="pink"
  >
    <svg
      aria-hidden="true"
      className="ColorIndicator__CheckIcon-sc-1w5zl7g-0 Lufbl StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
        fill="currentColor"
      />
    </svg>
    <input
      checked={false}
      name="pink"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="ColorIndicator__FocusedStyle-sc-1w5zl7g-1 bTsnCP"
      style={
        Object {
          "focused": "#de017e",
          "label": "pink",
          "passive": "#ffc5e1",
        }
      }
    />
  </label>
</div>
`;
